src/app/app.component.ts

```diff
@@ -0,0 +1,57 @@
+import { Component, OnInit } from '@angular/core';
+import { CoffeeShopsService } from './services/coffeeShops.service';
+import {  retry, switchMap } from 'rxjs';
+import { CoffeeShopModel } from './models/coffeeShop.model';
+import { FormControl, FormGroup, Validators } from '@angular/forms';
+
+@Component({
+  selector: 'app-root',
+  templateUrl: './app.component.html',
+  styleUrls: ['./app.component.scss']
+})
+export class AppComponent implements OnInit {
+  coffeeShops = new Array<CoffeeShopModel>();
+  coordinatesFormControl = new FormGroup({
+    xCoordinate: new FormControl([], Validators.required),
+    yCoordinate: new FormControl([], Validators.required)
+  });
+
+  constructor(private coffeeShop: CoffeeShopsService) {
+  }
+
+  ngOnInit(): void {
+    this.getCoffeeShops();
+  }
+
+  getCoffeeShops(): void {
+    this.coffeeShop.getToken().pipe(
```

Comment:
Before the retry(5), i tried with a retryWhen(errors => if(errors) return true ), but it wouldn't work for some reason. It could've been done with a simple switch, to check the error message and then to retry.